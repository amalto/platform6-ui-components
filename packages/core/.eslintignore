dist/
loader/

src/components.d.ts