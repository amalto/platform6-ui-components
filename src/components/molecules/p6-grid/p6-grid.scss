@import "bulma/sass/elements/button";
@import "bulma/sass/elements/tag";

:host {
  display: flex;
  flex-direction: column;

  .btn-bar {
    padding-bottom: $gap-size * 1px;
    width: 100%;

    p6-button {
      margin-right: $gap-size * 1px;
    }
  }

  .hidden-column-panel {
    border: 1px solid $grey-lighter;
    padding: $gap-size * 1px;
    margin-bottom: $gap-size * 1px;

    p6-tag {
      border-radius: 0;
      cursor: pointer;
      margin-right: $gap-size * 1px;
    }
  }

  .row-context-menu {
    display: none;
    position: absolute;
    z-index: 100;

    &.is-open {
      background-color: $white;
      border: 1px solid $white-ter;
      display: flex;
      padding: $gap-size * 1px;
    }
  }
}
