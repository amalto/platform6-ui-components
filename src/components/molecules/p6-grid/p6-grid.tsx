import {
  Component,
  Element,
  h,
  Host,
  Listen,
  Prop,
  State,
} from "@stencil/core";
import { Align, Direction, Operation } from "~shared/types";
import {
  getCellLabelByHeaderId,
  getRowCellByHeaderId,
  HeaderCell,
  RowCell,
  toggleSort,
} from "./utils";

export declare type EventCallBack = (event: MouseEvent) => void;

const MIN_WIDTH = 75;
const DEFAULT_WIDTH = MIN_WIDTH;
const INC_WIDTH = 10;
const DEFAULT_COL = "col-1";

@Component({
  tag: "p6-grid",
  styleUrl: "p6-grid.scss",
  shadow: true,
})
export class P6Tables {
  @Element() host!: HTMLP6GridElement;

  /**
   * Grid headers
   */
  @Prop() headers!: HeaderCell[];

  /**
   * Grid rows
   */
  @Prop() rows!: RowCell[][];

  /**
   * Update callback after each action on the grid
   */
  @Prop() updateGridCallback!: (
    header: HeaderCell[],
    rows: RowCell[][]
  ) => void;

  @State() sortedBy = DEFAULT_COL;

  @State() stateHeaders: HeaderCell[] = [];

  @State() stateRows: RowCell[][] = [];

  private clearTimeout: NodeJS.Timeout | undefined;

  private align = (id: string, align: Align): void => {
    this.updateGridCallback(
      this.updateHeaderAttr(id, "align", align),
      this.rows
    );
  };

  /**
   * Workaround until we find out why the listner is triggered 2 times.
   * @param method
   */
  private triggerOnce(method: () => void): void {
    if (!this.clearTimeout) {
      this.clearTimeout = setTimeout(() => {
        method();
        clearTimeout(this.clearTimeout as NodeJS.Timeout);
        this.clearTimeout = undefined;
      }, 100);
    }
  }

  @Listen("alignLeft")
  private alignLeft(event: CustomEvent<string>): void {
    this.triggerOnce(() => this.align(event.detail, "start"));
  }

  @Listen("alignCenter")
  private alignCenter(event: CustomEvent<string>): void {
    this.triggerOnce(() => this.align(event.detail, "center"));
  }

  @Listen("alignRight")
  private alignRight(event: CustomEvent<string>): void {
    this.triggerOnce(() => this.align(event.detail, "end"));
  }

  private calcOrder(id: string, operation: Operation): void {
    if (!this.isHeaderUndefined(id)) {
      const header: HeaderCell = this.getHeaderById(id) as HeaderCell;
      const width: number = header.width || DEFAULT_WIDTH;
      const minWidth: number =
        width - INC_WIDTH < MIN_WIDTH ? MIN_WIDTH : width - INC_WIDTH;
      const newWidth: number =
        operation === "minus" ? minWidth : width + INC_WIDTH;
      this.updateGridCallback(
        this.updateHeaderAttr(id, "width", newWidth),
        this.rows
      );
    }
  }

  private getHeaderById(id: string): HeaderCell | undefined {
    return this.stateHeaders.find((header) => header.id === id);
  }

  private getHeaderExcept(id: string): HeaderCell[] {
    return this.stateHeaders.filter((header) => header.id !== id);
  }

  @Listen("hide")
  private hide(event: CustomEvent<string>): void {
    this.triggerOnce(() => {
      const id: string = event.detail;
      if (!this.isHeaderUndefined(id)) {
        const header: HeaderCell = this.getHeaderById(id) as HeaderCell;
        const updatedHeader: HeaderCell[] = this.getHeaderExcept(id);

        header.hidden = !header.hidden;
        updatedHeader.push(header);
        this.stateHeaders = updatedHeader;
        this.updateGridCallback(updatedHeader, this.rows);
      }
    });
  }

  private isHeaderUndefined(id: string): boolean {
    const isUndefined = !this.getHeaderById(id);

    if (isUndefined) {
      // eslint-disable-next-line no-console
      console.warn(`Header with id ${id} doesn't exist!`);
    }
    return isUndefined;
  }

  @Listen("minus")
  private minus(event: CustomEvent<string>): void {
    this.triggerOnce(() => this.calcOrder(event.detail, "minus"));
  }

  private move(id: string, direction: Direction): void {
    if (!this.isHeaderUndefined(id)) {
      const { stateHeaders } = this;
      const updatedHeaders: HeaderCell[] = [];
      const idx: number = stateHeaders.findIndex((header) => header.id === id);
      const isLeft: boolean = direction === "left";
      const isLast: boolean = idx === stateHeaders.length - 1;
      const isFirst: boolean = idx === 0;
      const hasEnoughtItems: boolean = stateHeaders.length > 1;
      const canMoveLeft: boolean = !isFirst || !isLeft;
      const canMoveRight: boolean = !isLast || isLeft;

      if (hasEnoughtItems && canMoveLeft && canMoveRight) {
        for (let i = 0; i < stateHeaders.length; i += 1) {
          if (isLeft) {
            if (i === idx) {
              updatedHeaders.push(stateHeaders[idx - 1]);
            } else if (i !== idx - 1) {
              updatedHeaders.push(stateHeaders[i]);
            } else {
              updatedHeaders.push(stateHeaders[idx]);
            }
          } else if (i === idx) {
            updatedHeaders.push(stateHeaders[idx + 1]);
          } else if (idx !== i - 1) {
            updatedHeaders.push(stateHeaders[i]);
          } else {
            updatedHeaders.push(stateHeaders[idx]);
          }
        }
        this.stateHeaders = updatedHeaders;
      }
    }
  }

  @Listen("moveLeft")
  private moveLeft(event: CustomEvent<string>): void {
    this.triggerOnce(() => this.move(event.detail, "left"));
  }

  @Listen("moveRight")
  private moveRight(event: CustomEvent<string>): void {
    this.triggerOnce(() => this.move(event.detail, "right"));
  }

  @Listen("plus")
  private plus(event: CustomEvent<string>): void {
    this.triggerOnce(() => this.calcOrder(event.detail, "plus"));
  }

  @Listen("sort")
  private sort(event: CustomEvent<string>): void {
    this.triggerOnce(() => {
      const id: string = event.detail;
      if (!this.isHeaderUndefined(id)) {
        const { stateHeaders } = this;
        const idx: number = stateHeaders.findIndex(
          (header) => header.id === id
        );
        const updatedHeaders: HeaderCell[] = [...stateHeaders];

        updatedHeaders[idx].sort = toggleSort(
          updatedHeaders[idx].sort || "asc"
        );
        this.stateHeaders = updatedHeaders;
        this.sortedBy = id;
        this.updateGridCallback(updatedHeaders, this.rows);
      }
    });
  }

  @Listen("setColor")
  private setColor(event: CustomEvent<string>): void {
    this.triggerOnce(() => {
      const id: string = event.detail;
      // eslint-disable-next-line no-console
      console.info("id => ", id);
      this.updateGridCallback(this.stateHeaders, this.rows);
    });
  }

  private updateHeaderAttr(
    id: string,
    attrName: string,
    attr: unknown
  ): HeaderCell[] {
    const { stateHeaders } = this;
    this.stateHeaders = stateHeaders.map((header) => {
      if (header.id === id) {
        return {
          ...header,
          [attrName]: attr,
        };
      }
      return header;
    });
    return this.stateHeaders;
  }

  private renderHeader(): HTMLP6GridHeaderElement {
    const { stateHeaders } = this;
    const lastIndex: number = stateHeaders.length;
    const displayableHeader: HeaderCell[] = stateHeaders.filter(
      (cell) => !cell.hidden
    );
    return (
      <p6-grid-header>
        {displayableHeader.map((header, idx) => (
          <p6-grid-cell
            // eslint-disable-next-line react/jsx-no-bind
            onAlignLeft={this.alignLeft.bind(this)}
            // eslint-disable-next-line react/jsx-no-bind
            onAlignCenter={this.alignCenter.bind(this)}
            // eslint-disable-next-line react/jsx-no-bind
            onAlignRight={this.alignRight.bind(this)}
            color={header.color}
            headerId={header.id}
            // eslint-disable-next-line react/jsx-no-bind
            onHide={this.hide.bind(this)}
            // eslint-disable-next-line react/jsx-no-bind
            onMinus={lastIndex === idx ? undefined : this.minus.bind(this)}
            // eslint-disable-next-line react/jsx-no-bind
            onMoveLeft={idx === 0 ? undefined : this.moveLeft.bind(this)}
            // eslint-disable-next-line react/jsx-no-bind
            onMoveRight={
              lastIndex === idx ? undefined : this.moveRight.bind(this)
            }
            // eslint-disable-next-line react/jsx-no-bind
            onPlus={lastIndex === idx ? undefined : this.plus.bind(this)}
            // eslint-disable-next-line react/jsx-no-bind
            onSort={this.sort.bind(this)}
            // eslint-disable-next-line react/jsx-no-bind
            onSetColor={this.setColor.bind(this)}
            width={header.width}
          >
            {header.label}
          </p6-grid-cell>
        ))}
      </p6-grid-header>
    );
  }

  private renderRowCell = (cell: RowCell): HTMLP6GridCellElement => {
    const header: HeaderCell = this.getHeaderById(cell.headerId) as HeaderCell;
    return (
      <p6-grid-cell
        align={header.align}
        color={header.color}
        headerId={cell.headerId}
        width={header.width}
      >
        {cell.label}
      </p6-grid-cell>
    );
  };

  private renderRow = (
    cells: RowCell[],
    orderedHeaders: string[]
  ): HTMLP6GridRowElement => {
    const orderedCells: RowCell[] = [];

    orderedHeaders.forEach((id) => {
      orderedCells.push(cells.find((cell) => cell.headerId === id) as RowCell);
    });

    return <p6-grid-row>{orderedCells.map(this.renderRowCell)}</p6-grid-row>;
  };

  private renderRows = (): HTMLP6GridBodyElement | null => {
    const { rows, stateHeaders } = this;

    if (rows.length === 0) {
      return null;
    }

    const orderedHeaders: string[] = stateHeaders
      .filter((header) => !header.hidden)
      .map((header) => header.id);
    const sortedCellds: RowCell[][] = this.sortRows();

    // console.info("sortedCellds => ", sortedCellds);
    const rowsElements: HTMLP6GridRowElement[] = sortedCellds.map((row) => {
      return this.renderRow(row, orderedHeaders);
    });

    return <p6-grid-body>{rowsElements}</p6-grid-body>;
  };

  private sortRows(): RowCell[][] {
    const { sortedBy, stateRows } = this;
    const header: HeaderCell = this.getHeaderById(sortedBy) as HeaderCell;

    return stateRows.sort((a, b) => {
      const cellA: RowCell = getRowCellByHeaderId(sortedBy, a) as RowCell;
      const cellB: RowCell = getRowCellByHeaderId(sortedBy, b) as RowCell;
      const vA: string = getCellLabelByHeaderId(
        cellA.headerId,
        sortedBy,
        cellA.label
      );
      const vB: string = getCellLabelByHeaderId(
        cellB.headerId,
        sortedBy,
        cellB.label
      );

      return header.sort === "desc"
        ? vB.localeCompare(vA)
        : vA.localeCompare(vB);
    });
  }

  componentWillLoad(): void {
    this.stateHeaders = this.headers;
    this.stateRows = this.rows;
  }

  render(): JSX.Element {
    return (
      <Host>
        <div>
          <p6-button mode="danger" size="small" type="button">
            <p6-icon name="eraser" />
          </p6-button>
        </div>
        {this.renderHeader()}
        {this.renderRows()}
        <slot />
      </Host>
    );
  }
}
