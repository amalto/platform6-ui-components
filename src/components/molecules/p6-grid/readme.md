# p6-grid

<!-- Auto Generated Below -->

## Properties

| Property               | Attribute | Description                              | Type                                         | Default     |
| ---------------------- | --------- | ---------------------------------------- | -------------------------------------------- | ----------- |
| `customContextMenu`    | --        | Display a context menu based on row data | `((row: RowCell[]) => Element) \| undefined` | `undefined` |
| `headers` _(required)_ | --        | Grid headers                             | `HeaderCell[]`                               | `undefined` |
| `loading`              | `loading` | Display spinner                          | `boolean`                                    | `false`     |
| `rows` _(required)_    | --        | Grid rows                                | `Row[]`                                      | `undefined` |

## Events

| Event      | Description                                        | Type               |
| ---------- | -------------------------------------------------- | ------------------ |
| `p6Change` | Listen to change event to get updated p6-grid data | `CustomEvent<any>` |

## Dependencies

### Depends on

- [p6-grid-cell](./components/p6-grid-cell)
- [p6-grid-row](./components/p6-grid-row)
- [p6-grid-body](./components/p6-grid-body)
- [p6-spinner](../../atoms/p6-spinner)
- [p6-grid-header](./components/p6-grid-header)
- [p6-action](../../atoms/p6-action)
- [p6-icon](../../atoms/p6-icon)
- [p6-empty](../../templates/p6-empty)
- [p6-tag](../../atoms/p6-tag)

### Graph

```mermaid
graph TD;
  p6-grid --> p6-grid-cell
  p6-grid --> p6-grid-row
  p6-grid --> p6-grid-body
  p6-grid --> p6-spinner
  p6-grid --> p6-grid-header
  p6-grid --> p6-action
  p6-grid --> p6-icon
  p6-grid --> p6-empty
  p6-grid --> p6-tag
  p6-grid-cell --> p6-action
  p6-grid-cell --> p6-icon
  style p6-grid fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
