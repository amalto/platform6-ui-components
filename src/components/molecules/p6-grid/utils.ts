import { isEmpty } from "~utils/attribute";

const HEADER_DATA_TYPE = "header-cell";
const ROW_DATA_TYPE = "row-cell";

const DATA_ID = "data-id";
const DATA_TYPE = "data-type";

export interface Row extends HTMLElement {
  "data-id": string;
  "data-type": string;
}

export function isAttrInElement(element: Element, attrName: string): boolean {
  return !isEmpty(element.getAttribute(attrName));
}

export function checkDataType(
  element: Element,
  expectedDataType: string
): boolean {
  return element.getAttribute(DATA_TYPE) === expectedDataType;
}

export function isValidElement(element: Element, attrs: string[]): boolean {
  return !attrs.some((attr) => !isAttrInElement(element, attr));
}

export function isValidCell(
  element: Element,
  expectedDataType: string
): boolean {
  return (
    isValidElement(element, [DATA_ID, DATA_TYPE]) &&
    checkDataType(element, expectedDataType)
  );
}

export function isHeaderCell(element: Element): element is Row {
  return isValidCell(element, HEADER_DATA_TYPE);
}

export function isRowCell(element: Element): element is Row {
  return isValidCell(element, ROW_DATA_TYPE);
}
