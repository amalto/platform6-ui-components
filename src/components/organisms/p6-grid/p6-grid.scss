:host {
  display: flex;
  flex-direction: column;

  .row-context-menu {
    display: none;
    position: absolute;
    z-index: 100;

    &.is-open {
      background-color: $white;
      border: 1px solid $white-ter;
      display: flex;
      padding: $gap-size * 1px;
    }
  }

  p6-grid-actions + :not(p6-grid-header) {
    margin-top: $gap-size * 1px;
  }

  p6-grid-header + :not(p6-grid-body) {
    margin-top: $gap-size * 0.21rem;
  }
}
