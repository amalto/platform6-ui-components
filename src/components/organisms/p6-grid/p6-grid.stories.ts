import { Alignment } from "../../../shared/types";
import { ColumnDefinition, DataItem, Row } from "./core/entities";

export default {
  title: "Organisms/Grid",
  component: "p6-grid",
};

interface StoryItem {
  first: string;
  second: string;
  third: string;
  [key: string]: string;
}

const definitions: ColumnDefinition<DataItem>[] = [
  {
    id: "first",
    label: "First Col",
    field: "first",
    editable: true,
    width: 150,
  },
  {
    label: "Second Col",
    field: "second",
    editable: true,
    width: 150,
    align: Alignment.End,
  },
  {
    id: "third",
    label: "Third Col",
    field: "third",
    editable: true,
    getValue: (data) => data.third as string,
  },
  {
    label: "Fourth",
    editable: false,
    hidden: true,
    disableHeaderMenu: true,
    getValue: (data) => (data.fourth as string) || "N/A",
  },
  {
    label: "Last",
    editable: false,
    width: 300,
    getValue: (data) => Object.values(data).join("-"),
  },
];

const storyItems: StoryItem[] = [
  {
    first: "row(1) - col(first)",
    second: "row(1) - col(second)",
    third: "row(1) - col(third)",
  },
  {
    first: "row(2) - col(first)",
    second: "row(2) - col(second)",
    third: "row(2) - col(third)",
    fourth: "row(2) - col(Fourth)",
  },
  {
    first: "row(3) - col(first)",
    second: "row(3) - col(second)",
    third: "row(3) - col(third)",
  },
  {
    first: "row(4) - col(first)",
    second: "row(4) - col(second)",
    third: "row(4) - col(third)",
  },
  {
    first: "row(5) - col(first)",
    second: "row(5) - col(second)",
    third: "row(5) - col(third)",
  },
  {
    first: "row(6) - col(first)",
    second: "row(6) - col(second)",
    third: "row(6) - col(third)",
    fourth: "row(6) - col(Fourth)",
    fifth: "row(6) - col(Fourth)",
  },
];

const customContextMenu = (row: Row<DataItem>): JSX.Element => {
  const data = Object.entries(row.data)
    .map(([key, value]) => `â‡¾ ${key}: ${value}`)
    .join("\n");

  return (`Context menu exemple, ${row.id} : ${data}` as unknown) as JSX.Element;
};

export const Sample = (): HTMLP6GridElement => {
  const kLoading = boolean("Loading", false);
  const kEmpty = boolean("Empty", false);

  const grid = document.createElement("p6-grid");
  grid.loading = kLoading;
  grid.definitions = definitions;
  grid.data = kEmpty ? [] : storyItems;
  grid.loading = kLoading;
  grid.customContextMenu = customContextMenu;

  return grid;
};

export const Loading = (): HTMLP6GridElement => {
  const grid = document.createElement("p6-grid");
  grid.loading = true;
  grid.definitions = definitions;
  grid.data = storyItems;
  return grid;
};

export const Empty = (): HTMLP6GridElement => {
  const grid = document.createElement("p6-grid");
  grid.definitions = definitions;
  grid.data = [];
  return grid;
};
