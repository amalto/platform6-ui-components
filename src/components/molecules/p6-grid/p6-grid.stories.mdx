import { Story, Preview, Meta } from "@storybook/addon-docs/blocks";
import { action } from "@storybook/addon-actions";
import readme from "./readme.md";

<Meta
  title="Molecules/Grid"
  component="p6-grid"
  parameters={{ notes: readme }}
/>

# Grid

`p6-grid` provides a grid component with templating.

<Preview>
  <Story name="Default" height="300px">
    {() => {
      let headers = [
        { color: "blue", id: "col-1", label: "First", width: 150 },
        { id: "col-2", label: "Second", width: 100 },
        { id: "col-3", label: "Third", width: 100 },
        { id: "col-4", label: "Fourth", width: 100 },
      ];
      let rows = [
        {
          cells: [
            { headerId: "col-2", label: "1 second cell row 1" },
            {
              align: "center",
              color: "blue",
              headerId: "col-1",
              label: "1 first cell row 1",
              width: 150,
            },
            { headerId: "col-3", label: "1 third cell row 1" },
            { headerId: "col-4", label: "1 fourth cell row 1" },
          ],
        },
        {
          cells: [
            { headerId: "col-2", label: "2 second cell row 2" },
            { headerId: "col-3", label: "2 third cell row 2" },
            {
              align: "center",
              color: "blue",
              headerId: "col-1",
              label: "2 first cell row 2",
              width: 150,
            },
            { headerId: "col-4", label: "2 fourth cell row 2" },
          ],
        },
        {
          cells: [
            { headerId: "col-2", label: "3 second cell row 3" },
            { headerId: "col-3", label: "3 third cell row 3" },
            { headerId: "col-4", label: "3 fourth cell row 3" },
            {
              align: "center",
              color: "blue",
              headerId: "col-1",
              label: "3 first cell row 3",
              width: 150,
            },
          ],
        },
      ];
      const grid = document.createElement("p6-grid");
      const body = document.createElement("p6-grid-body");
      grid.id = "default-grid";
      grid.lang = "fr";
      grid.headers = headers;
      grid.rows = rows;
      grid.updateGridCallback = (updatedHeaders, updatedRows) => {
        {
          /* headers = updatedHeaders;
        rows = updatedRows; */
        }
      };
      return grid;
    }}
  </Story>
</Preview>
