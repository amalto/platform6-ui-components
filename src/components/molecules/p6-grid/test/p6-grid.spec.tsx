import { newSpecPage } from "@stencil/core/testing";
import { P6Tables } from "../p6-grid";

// https://stackoverflow.com/questions/52868727/how-to-mock-window-navigator-language-using-jest
describe("p6-grid", () => {
  it("renders", async () => {
    const page = await newSpecPage({
      components: [P6Tables],
      html: `<p6-grid></p6-grid>`,
    });
    expect(page.root).toEqualHtml(`
      <p6-grid>
        <mock:shadow-root>
          <slot></slot>
        </mock:shadow-root>
      </p6-grid>
    `);
  });
});
