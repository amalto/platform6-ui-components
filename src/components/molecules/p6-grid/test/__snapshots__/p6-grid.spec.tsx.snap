// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-grid should render an empty grid 1`] = `
<p6-grid>
  <mock:shadow-root></mock:shadow-root>
</p6-grid>
`;

exports[`p6-grid should render a grid with on row 1`] = `
<p6-grid id="p6-grid-test">
  <mock:shadow-root></mock:shadow-root>
</p6-grid>
`;